Implement "Create" Instrument Listing Feature

moelgazzar opened the pull request:
This feature creates new instrument listings, which can be accessed at **/instruments/new**

- Added "new" and "create" actions in InstrumentsController.
- Created "new.html.erb" view for the instrument form "using simple_form".
- Updated the Instrument model with validations for required fields (name, address, category, price, description).
- Modified the routes to include routes for "new" and "create" actions.

moelgazzar committed:
Add routes for instrument new and create actions

moelgazzar committed:
Implement new and create actions in instruments_controller

moelgazzar committed:
Create new instrument form using simple_form

moelgazzar committed:
Add validations to Instrument model, and updated the /index and /show pages to be in line with the /new page

p-guelfi commented:
Looks good!

DavidDicke commented on app/controllers/instruments_controller.rb:
```diff
@@ -1,9 +1,35 @@
 class InstrumentsController < ApplicationController
+  # This ensures that users are logged in for actions that modify data
+  before_action :authenticate_user!, only: [:new, :create]
```
I think in the lecture they taught that the best practice would be the white listing approach. So protect every route in the applications controller as a default and then we would not have to insert it here, but do `skip_before_action :authenticate_user!, only: %i[index home]`

DavidDicke commented on app/controllers/instruments_controller.rb:
```diff
@@ -1,9 +1,35 @@
+    if @instrument.save
+      redirect_to instrument_path(@instrument), notice: 'Instrument successfully created!'
+    else
+      render :new, alert: 'There was an issue creating your instrument listing.'
```
We should add the `status: :unprocessable_entity`

DavidDicke commented on app/models/instrument.rb:
```diff
@@ -1,3 +1,9 @@
 class Instrument < ApplicationRecord
   belongs_to :user
+
```
we can add the `has_many_attached :photos` here as well.

DavidDicke commented on app/models/instrument.rb:
```diff
@@ -1,3 +1,9 @@
+  validates :description, presence: true
+  validates :category, presence: true
+  validates :address, presence: true
+  validates :price, numericality: { greater_than_or_equal_to: 0 }
```
also: `presence: true`

DavidDicke commented on app/views/instruments/new.html.erb:
```diff
@@ -0,0 +1,9 @@
+  <%= f.input :address, label: "Address" %>
+  <%= f.input :price, label: "Price" %>
+  <%= f.input :category, label: "Category" %>
+
```
If we have the attached photos in model, we should have an upload field here: `<%= f.input :photos, as: :file, input_html: { multiple: true } %>`

DavidDicke commented on app/views/instruments/show.html.erb:
```diff
@@ -3,6 +3,8 @@
 
```
Here we would iterate over the photos and show them with the cl_image_tag maybe as a carousel.

DavidDicke reviewed (commented):
Hey I added some comments regarding best practices and the Photos that we want to add to instruments.

moelgazzar committed:
Add has_many_attached :photos to Instrument model

moelgazzar committed:
Update instrument form to support photo uploads

moelgazzar committed:
Display uploaded photos in the instrument show view

moelgazzar committed:
Refactor instruments controller to centralize authentication and improve error handling

moelgazzar commented:
Hi @DavidDicke, @p-guelfi,

I have addressed all the feedback and made the following adjustments:

- Centralized authentication by adding `before_action :authenticate_user!` in `ApplicationController`.
- Whitelisted the `index` and `show` actions in `InstrumentsController` using `skip_before_action :authenticate_user!`.
- Updated the `create` action in `InstrumentsController`:
  - Added `flash[:alert]` to show an error message when form submission fails.
  - Used `status: :unprocessable_entity` to give the correct HTTP status for invalid form submissions.
- Updated the `new.html.erb` view to display `flash[:alert]` messages when an issue occurs during instrument creation.

Let me know if there’s anything else to adjust.

Thanks!

moelgazzar committed:
Merge branch 'master' into instrument-listing-page

# Conflicts:
#	app/controllers/instruments_controller.rb
#	app/models/instrument.rb
#	app/views/instruments/show.html.erb
#	config/routes.rb
